### PR TITLE
[var_grad] Add var_grad for input/output lists

### PR DESCRIPTION
**Pull request by kparichay**
Added var_grad for input/output lists which also combines derivatives
    This is the baseclass for the weights
    This update will help with graph class
    
**Self evaluation:**
    1. Build test: [x]Passed [ ]Failed [ ]Skipped
    2. Run test: [x]Passed [ ]Failed [ ]Skipped
    
Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -37,8 +37,8 @@ warning_flags = [
   '-Wvla',
   '-Wpointer-arith',
   '-Wno-error=varargs',
-  '-O2',
-  '-ftree-vectorize'
+  #  '-O2',
```
Comment (zhoonit) [inline meson.build:40]: Maybe unintended comment?

File: nntrainer/tensor/var_grad.h [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+  Var_Grad() : trainable(false) {}
+
+  /**
+   * @brief Construct a new Var_Grad object
+   *
+   * @param dim Variable and gradient tensor dimension
+   * @param train If the variable is trainable
+   * @param name Name for this Var_Grad
+   */
+  Var_Grad(const TensorDim &dim, bool train = true, std::string name = "");
```
Comment (zhoonit) [inline nntrainer/tensor/var_grad.h:40]: Maybe `const std::string & name`?

File: nntrainer/tensor/var_grad.h [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+  /**
+   * @brief Swap for Var_Grad
+   *
+   * @param lhs Swap to
+   * @param rhs Swap from
+   * @note Only swap gradient if trainable
```
Comment (zhoonit) [inline nntrainer/tensor/var_grad.h:47]: I guess this note should go away

**Review by zhoonit — COMMENTED** [review]
Could I get an explanation for the `var_grad`? I don't see the usage in the code and can't think of a reason to factor out those

File: meson.build (hunk 1 above)
Comment (kparichay) [reply meson.build]: done.

File: nntrainer/tensor/var_grad.h (hunk 2 above)
Comment (kparichay) [reply nntrainer/tensor/var_grad.h]: done.

File: nntrainer/tensor/var_grad.h (hunk 3 above)
Comment (kparichay) [reply nntrainer/tensor/var_grad.h]: done.

**Comment by kparichay** [comment]
> Could I get an explanation for the `var_grad`? I don't see the usage in the code and can't think of a reason to factor out those

`var_grad` will replace inputs, outputs, derivative, ret_derivative for layers.

**Review by zhoonit — APPROVED** [review]
LGTM

File: test/unittest/unittest_nntrainer_layers.cpp [hunk 4]
```diff
@@ -1034,16 +1034,13 @@ TEST_F(nntrainer_Conv2DLayer, save_read_01_p) {
...
-
-  read_file.seekg(0, std::ios::beg);
 
   std::ifstream read_file2("save1.bin");
   ASSERT_TRUE(read_file2.good());
 
   float d1, d2;
 
-  for (int i = 0; i < (5 * 5 * 3 * 6) + 6; ++i) {
+  for (int i = 0; i < (5 * 5 * 3 * 12) + 12; ++i) {
```
Comment (lhs8928) [inline test/unittest/unittest_nntrainer_layers.cpp:1043]: Does this change is related to var_grad?

Comment (kparichay) [reply test/unittest/unittest_nntrainer_layers.cpp:1043]: It does not directly. The test in this case was not being performed on the full weight which was revealed when using this patch.

**Comment by lhs8928** [comment]
LGTM

**Review by lhs8928 — APPROVED** [review]

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Pull request merged**